### ISSUE TITLE
Occasional deadlooks freeze ST on Windows

### ISSUE THREAD

**Issue by deathaxe** [tomv564/LSP#12]
I recently have experienced ST to freeze with little CPU usage while working in a python file which is attached to python-language-server 0.4.1 using LSP package.

All other language supporting packages are disabled after installing LSP some days ago.

It looks like there are some deadlocks somewhere. Situation seemed to become worse, when I tried to disable debug logging which pollutes the console.

Not quite sure how to find out how it is triggered - maybe it is caused by on_hover event.

Windows 10 x64 / ST 4142 x64

**Comment by tomv564** [comment]
Thanks for the bug report, and extra thanks for helping investigate!
This must be the recent ST dev build, 3142, which _should_ be more performant, not less.

If you can still have the debug log on, a chunk of log around the freezes would be helpful.

The `on_hover` event calls Jedi via pyls, and does markdown formatting. Are you able to test with the `on_hover` handler commented out?

LSP calls `textDocument/didChange` 500ms after the last edit or before requesting completion, there could be issues from typing with short pauses. Do you notice the freezes while editing?

**Comment by deathaxe** [comment]
I'll keep an eye on it and post any possibly helpful information. I already tried to reproduce the freezing yesterday, but as soon as the console is open with logging enabled ST seems not to freeze any more. Don't have any glue why right now.

**Comment by deathaxe** [comment]
I realized an ST freeze once now after LSP failed to decode the output from python language server. A second try with same error did not cause a freeze. Strange. May the uncaught `UnicodeDecodeError` cause some mutex/lock objects to stay there forever, sometimes?

```
LSP: {'contents': 'on_hover(self, view, point, hover_zone)\n\nHandle the hover event and show tooltip if needed.'} 
LSP: (stderr):  2017-08-18 16:28:30,377 UTC - ERROR - pyls.language_server - CAUGHT 
LSP: (stderr):  Traceback (most recent call last): 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\pyls\language_server.py", line 67, in wrapped 
LSP: (stderr):  return func(*args, **kwargs) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\pyls\python_ls.py", line 137, in m_text_document__hover 
LSP: (stderr):  return self.hover(textDocument['uri'], position) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\pyls\python_ls.py", line 88, in hover 
LSP: (stderr):  return self._hook(self._hooks.pyls_hover, doc_uri, position=position) or {'contents': ''} 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\pyls\python_ls.py", line 57, in _hook 
LSP: (stderr):  return hook(config=self.config, workspace=self.workspace, document=doc, **kwargs) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\pluggy.py", line 745, in __call__ 
LSP: (stderr):  return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\pluggy.py", line 339, in _hookexec 
LSP: (stderr):  return self._inner_hookexec(hook, methods, kwargs) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\pluggy.py", line 302, in __call__ 
LSP: (stderr):  return outcome.get_result() 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\pluggy.py", line 279, in get_result 
LSP: (stderr):  raise ex[1].with_traceback(ex[2]) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\pluggy.py", line 265, in __init__ 
LSP: (stderr):  self.result = func() 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\pluggy.py", line 300, in <lambda> 
LSP: (stderr):  outcome = _CallOutcome(lambda: self.oldcall(hook, hook_impls, kwargs)) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\pluggy.py", line 334, in <lambda> 
LSP: (stderr):  _MultiCall(methods, kwargs, hook.spec_opts).execute() 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\pluggy.py", line 614, in execute 
LSP: (stderr):  res = hook_impl.function(*args) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\pyls\plugins\hover.py", line 10, in pyls_hover 
LSP: (stderr):  definitions = document.jedi_script(position).goto_definitions() 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\api\__init__.py", line 203, in goto_definitions 
LSP: (stderr):  definitions = helpers.evaluate_goto_definition(self._evaluator, context, leaf) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\api\helpers.py", line 198, in evaluate_goto_definition 
LSP: (stderr):  return evaluator.goto_definitions(context, leaf) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\__init__.py", line 462, in goto_definitions 
LSP: (stderr):  return helpers.evaluate_call_of_leaf(context, name) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\helpers.py", line 53, in evaluate_call_of_leaf 
LSP: (stderr):  return context.eval_node(leaf) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\context.py", line 37, in eval_node 
LSP: (stderr):  return self.evaluator.eval_element(self, node) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\__init__.py", line 247, in eval_element 
LSP: (stderr):  return self._eval_element_if_evaluated(context, element) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\__init__.py", line 264, in _eval_element_if_evaluated 
LSP: (stderr):  return self._eval_element_cached(context, element) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\cache.py", line 41, in wrapper 
LSP: (stderr):  rv = function(obj, *args, **kwargs) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\__init__.py", line 268, in _eval_element_cached 
LSP: (stderr):  return self._eval_element_not_cached(context, element) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\debug.py", line 81, in wrapper 
LSP: (stderr):  return func(*args, **kwargs) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\__init__.py", line 276, in _eval_element_not_cached 
LSP: (stderr):  types = self.eval_atom(context, element) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\__init__.py", line 347, in eval_atom 
LSP: (stderr):  search_global=True 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\_compatibility.py", line 213, in <lambda> 
LSP: (stderr):  return lambda *args, **kwargs: self.func(obj, *args, **kwargs) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\context.py", line 52, in py__getattribute__ 
LSP: (stderr):  self, name_or_str, name_context, position, search_global, is_goto) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\__init__.py", line 130, in find_types 
LSP: (stderr):  return f.find(filters, attribute_lookup=not search_global) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\debug.py", line 81, in wrapper 
LSP: (stderr):  return func(*args, **kwargs) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\finder.py", line 64, in find 
LSP: (stderr):  types = self._names_to_types(names, attribute_lookup) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\finder.py", line 147, in _names_to_types 
LSP: (stderr):  types = unite(name.infer() for name in names) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\common.py", line 191, in unite 
LSP: (stderr):  return set(typ for types in iterable for typ in types) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\common.py", line 191, in <genexpr> 
LSP: (stderr):  return set(typ for types in iterable for typ in types) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\finder.py", line 147, in <genexpr> 
LSP: (stderr):  types = unite(name.infer() for name in names) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\filters.py", line 117, in infer 
LSP: (stderr):  return self.get_param().infer() 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\filters.py", line 104, in get_param 
LSP: (stderr):  params = self.parent_context.get_params() 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\cache.py", line 41, in wrapper 
LSP: (stderr):  rv = function(obj, *args, **kwargs) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\representation.py", line 396, in get_params 
LSP: (stderr):  return param.get_params(self.evaluator, self.parent_context, self.tree_node, self.var_args) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\param.py", line 240, in get_params 
LSP: (stderr):  unpacked_va = list(var_args.unpack(func)) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\instance.py", line 421, in unpack 
LSP: (stderr):  for values in self._get_var_args().unpack(func): 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\instance.py", line 394, in unpack 
LSP: (stderr):  self._funcdef 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\debug.py", line 81, in wrapper 
LSP: (stderr):  return func(*args, **kwargs) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\dynamic.py", line 78, in search_params 
LSP: (stderr):  funcdef 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\cache.py", line 41, in wrapper 
LSP: (stderr):  rv = function(obj, *args, **kwargs) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\common.py", line 196, in wrapper 
LSP: (stderr):  return list(func(*args, **kwargs)) 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\dynamic.py", line 114, in _search_function_executions 
LSP: (stderr):  evaluator, [module_context], func_string_name): 
LSP: (stderr):  File "c:\apps\python36\lib\site-packages\jedi\evaluate\imports.py", line 530, in get_modules_containing_name 
LSP: (stderr):  for file_name in os.listdir(d): 
Exception in thread Thread-39:
Traceback (most recent call last):
  File "./python3.3/threading.py", line 901, in _bootstrap_inner
  File "./python3.3/threading.py", line 858, in run
  File "C:\Apps\Sublime\Data\Packages\LSP\main.py", line 376, in read_stderr
    error = self.process.stderr.readline().decode('UTF-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfc in position 36: invalid start byte
```

**Comment by deathaxe** [comment]
I modified the debug function to write log output to a file, closed console and opened a file. 

```python
def debug(*args):
    """Print args to the console if the "debug" setting is True."""
    # if settings.get('debug'):
    printf(*args)
    with open('C:\\data\\lsp.log', 'a') as file:
        print(*args, file=file, flush=True)
```

After some seconds and little modification ST freezes. Here is the content of the lsp.log so far.

```
starting in C:\
starting ['pyls']
client registered for window 2 {'pyls': <LSP.main.Client object at 0x00000296183730B8>}
got json:  {"result":{"capabilities":{"codeActionProvider":true,"codeLensProvider":{"resolveProvider":false},"completionProvider":{"resolveProvider":false,"triggerCharacters":["."]},"documentFormattingProvider":
notify: textDocument/didOpen
got json:  {"method":"textDocument/publishDiagnostics","params":{"uri":"file:///C:/Apps/Sublime/Data/Packages/LSP/main.py","diagnostics":[{"source":"pycodestyle","range":{"start":{"line":615,"character":79},"end
notify: textDocument/didOpen
got json:  {"method":"textDocument/publishDiagnostics","params":{"uri":"file:///C:/Apps/Sublime/Data/Packages/CNC%20Sinumerik%20840D/gcode/events.py","diagnostics":[{"source":"pycodestyle","range":{"start":{"lin
got json:  {"result":{"contents":"applies_to_primary_view_only(cls)\n\n"},"id":2,"jsonrpc":"2.0"}
{'contents': 'applies_to_primary_view_only(cls)\n\n'}
got json:  {"result":{"contents":""},"id":3,"jsonrpc":"2.0"}
{'contents': ''}
got json:  {"result":{"contents":""},"id":4,"jsonrpc":"2.0"}
{'contents': ''}
got json:  {"result":{"contents":""},"id":5,"jsonrpc":"2.0"}
{'contents': ''}
got json:  {"result":{"contents":"subscribe(cls, key, listener)\n\n"},"id":6,"jsonrpc":"2.0"}
{'contents': 'subscribe(cls, key, listener)\n\n'}
notify: textDocument/didChange
```

**Comment by deathaxe** [comment]
The screenshot shows another situation, which caused ST to freeze. I closed one view, which causes some events to be triggered quickly after each other. The closed tab did not yet completely disappear, when ST stops responding.

![freeze2](https://user-images.githubusercontent.com/16542113/29465426-8d0d9c40-8439-11e7-8a27-2a8cbf88e786.png)

**Comment by tomv564** [comment]
It looks like your python language server is not happy, it would be good to report some stack traces to https://github.com/palantir/python-language-server
```
Exception in thread Thread-39:
Traceback (most recent call last):
  File "./python3.3/threading.py", line 901, in _bootstrap_inner
  File "./python3.3/threading.py", line 858, in run
  File "C:\Apps\Sublime\Data\Packages\LSP\main.py", line 376, in read_stderr
    error = self.process.stderr.readline().decode('UTF-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfc in position 36: invalid start byte
```
This is definitely something for LSP to fix.

LSP should also not read stderr by default, this should be part of https://github.com/tomv564/LSP/issues/15

You can use pyls's `--log-file` or `--log-config` to capture its logging output without burdening LSP with it.

Your screenshot shows the same UnicodeDecodeError as earlier, I wonder if something happens to the language server's io streams during the DecodeError that causes other writes to hang?

Does the language server otherwise survive these freezes or does it restart?

**Comment by deathaxe** [comment]
Added `--log-file nul` to LSP's pyls config with following result.

After some hovering and editing, the log shows the following result, which I did not see before. As stderr is no longer sent to LSP package in that situation,  ST does not seem to freeze any longer, but the output shows some German umlauts and some escaped unicode characters, which may not have been handled correctly in stderr channel.

From this point on, pyls no longer works. 

The question now is, who causes the invalid paths with too many backslashes? Seems to be some escaping issue. Does LSP accidentally escape the backslashes too often or is it caused by pyls? 

```
LSP: got json:  {"error":{"code":-32000,"message":"Server error","data":{"type":"OSError","args":[22,"Die Syntax f\u00fcr den Dateinamen, Verzeichnisnamen oder die Datentr\u00e4gerbezeichnung ist falsch"],"message":" 
LSP: got error:  {'code': -32000, 'message': 'Server error', 'data': {'message': "[WinError 123] Die Syntax für den Dateinamen, Verzeichnisnamen oder die Datenträgerbezeichnung ist falsch: 'C:\\\\C:\\\\Apps\\\\Sublime\\\\Data\\\\Packages\\\\LSP'", 'args': [22, 'Die Syntax für den Dateinamen, Verzeichnisnamen oder die Datenträgerbezeichnung ist falsch'], 'type': 'OSError'}} 
LSP: got json:  {"result":{"contents":"on_hover(self, point, hover_zone)\n\n"},"id":14,"jsonrpc":"2.0"} 
LSP: {'contents': 'on_hover(self, point, hover_zone)\n\n'} 
LSP: got json:  {"error":{"code":-32000,"message":"Server error","data":{"type":"OSError","args":[22,"Die Syntax f\u00fcr den Dateinamen, Verzeichnisnamen oder die Datentr\u00e4gerbezeichnung ist falsch"],"message":" 
LSP: got error:  {'code': -32000, 'message': 'Server error', 'data': {'message': "[WinError 123] Die Syntax für den Dateinamen, Verzeichnisnamen oder die Datenträgerbezeichnung ist falsch: 'C:\\\\C:\\\\Apps\\\\Sublime\\\\Data\\\\Packages\\\\LSP'", 'args': [22, 'Die Syntax für den Dateinamen, Verzeichnisnamen oder die Datenträgerbezeichnung ist falsch'], 'type': 'OSError'}} 
LSP: got json:  {"result":{"contents":""},"id":16,"jsonrpc":"2.0"} 
LSP: {'contents': ''} 
```

**Comment by deathaxe** [comment]
Looks like pyls is hanging after the crash caused by faulty path translation. As a result LSP hangs while waiting for an answer. pyls causes 100% CPU usage of one core. As soon as pyls is killed ST continues working properly - except no more completions, etc. are available

**Comment by deathaxe** [comment]
The trouble is/was caused by python-language-server (see: https://github.com/palantir/python-language-server/issues/94 and https://github.com/palantir/python-language-server/issues/96)

Fixes are proposed here: https://github.com/palantir/python-language-server/pull/95

**Comment by rwols** [comment]
Awesome detective work @deathaxe , and even a fix 👍

**Comment by tomv564** [comment]
I removed the UTF-8 decoding in the stderr reader in the recently merged #34. 
Is it ok if we close this issue?
